src/main/java/org/mal_lang/langspec/LangReader.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2020-2021 Foreseeti AB <https://foreseeti.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.mal_lang.langspec;
+
+import static org.mal_lang.langspec.Utils.checkNotNull;
+
+import jakarta.json.Json;
+import jakarta.json.JsonException;
+import jakarta.json.JsonObject;
+import jakarta.json.stream.JsonParsingException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.zip.ZipInputStream;
+
+/** Reads {@link Lang} objects from {@code .mar} files. */
+public final class LangReader {
+  private final InputStream in;
+  private JsonObject langSpec = null;
+  private final Map<String, byte[]> icons = new LinkedHashMap<>();
+  private String license = null;
+  private String notice = null;
+
+  /**
+   * Constructs a new {@code LangReader} object.
+   *
+   * @param in an input stream of a {@code .mar} file
+   * @throws NullPointerException if {@code in} is {@code null}
+   */
+  public LangReader(InputStream in) {
+    checkNotNull(in);
+    this.in = in;
+  }
+
+  private void readLangSpec(ZipInputStream zipIn) throws IOException {
+    var jsonReaderFactory = Json.createReaderFactory(null);
+    var jsonReader = jsonReaderFactory.createReader(zipIn, StandardCharsets.UTF_8);
+    try {
+      this.langSpec = jsonReader.readObject();
+    } catch (JsonParsingException e) {
+      throw new IllegalArgumentException("Failed to parse \"langspec.json\" as a JSON object", e);
+    } catch (JsonException e) {
+      var cause = e.getCause();
+      if (cause == null || !(cause instanceof IOException)) {
+        throw new RuntimeException("JsonException did not contain an IOException cause");
+      }
+      throw (IOException) cause;
+    }
+  }
+
+  private void readIcon(ZipInputStream zipIn, String name) throws IOException {
+    // Remove "lang/" prefix
+    var fileName = name.substring(6);
+    if (!fileName.endsWith(".png") && !fileName.endsWith(".svg")) {
+      return;
```

Comment:
exception?

Reply:
I don't think it should be an error to have other files in the archive. This `return` just ignores files under `icons/` that don't end with `.png` or `.svg`. I just noticed the typo on line 68, I'll fix that.